README.md
# exporting hub
Microservice for exporting to the hub.

The service is based on the [pipe-connector](https://github.com/piveau-data/piveau-pipe-connector) library. Any configuration applicable for the pipe-connector can also be used for this service.

## Table of Contents
1. [Build](#build)
1. [Run](#run)
1. [Docker](#docker)
1. [Configuration](#configuration)
    1. [Pipe](#pipe)
    1. [Data Info Object](#data-info-object)
    1. [Logging](logging)
1. [License](#license)

## Build
Requirements:
 * Git
 * Maven
 * Java

```bash
$ git clone https://github.com/piveau-data/piveau-consus-exporting-hub.git
$ cd piveau-consus-exporting-hub
$ mvn package
```

## Run

```bash
$ java -jar target/piveau-exporting-hub-far.jar
```

## Docker

Build docker image:

```bash
$ docker build -t piveau/piveau-exporting-hub .
```

Run docker image:

```bash
$ docker run -it -p 8080:8080 piveau/piveau-exporting-hub
```

## Configuration

### Pipe

_optional_

* `hub`

    Configuration object for the hub
    
* `hub/serviceName`

    The service name of the hub

Example
```json
{
  "hub": {
    "serviceName": "piveau-hub"
  }
}
```

### Data Info Object
The service understands the following properties as part of the data info object.
* `content` 

    The type of content of the data, if different from dataset. Possible values are `indentifierList` and `metrics` (optional, if content of the data is a dataset)
    
    
* `total`
 
    Total number of datasets (optional)

* `counter` 

    The number of this dataset (optional)

* `identifier` 

    The unique identifier of this dataset from the source side (required, when `content`does _not_ equal `identifierList`)

* `catalogue`

    The id of the target catalogue (required)
    
* `hash` 

    The hash value calculated at the source (optional)

Example
```json
{
  "total": 20300,
  "counter": 5,
  "identifier": "dataset_1",
  "catalogue": "govdata",
  "hash": "hexvalue"
}
```

### Environment
See also [pipe-connector](https://github.com/piveau-data/piveau-pipe-connector)

| Variable| Description | Default Value |
| :--- | :--- | :--- |
| `PIVEAU_HUB_ADDRESS` | The hub address used when no endpoint is set in the pipe segment configuration | none |
| `PIVEAU_HUB_APIKEY` | The hub api key to use when no endpoint is set in the pipe segment configuration | none |
| `PIVEAU_HUB_ADD_HASH` | Generate a canonical hash value if no hash is present | false |

### Logging
See [logback](https://logback.qos.ch/documentation.html) documentation for more details

| Variable| Description | Default Value |
| :--- | :--- | :--- |
| `PIVEAU_PIPE_LOG_APPENDER` | Configures the log appender for the pipe context | `STDOUT` |
| `PIVEAU_LOGSTASH_HOST`            | The host of the logstash service | `logstash` |
| `PIVEAU_LOGSTASH_PORT`            | The port the logstash service is running | `5044` |
| `PIVEAU_PIPE_LOG_PATH`     | Path to the file for the file appender | `logs/piveau-pipe.%d{yyyy-MM-dd}.log` |
| `PIVEAU_PIPE_LOG_LEVEL`    | The log level for the pipe context | `INFO` |
| `PIVEAU_LOG_LEVEL`    | The general log level for the `io.piveau` package | `INFO` |

## License

[Apache License, Version 2.0](LICENSE.md)


CHANGELOG.md
# ChangeLog

## [1.0.0](https://gitlab.fokus.fraunhofer.de/viaduct/piveau-exporting-hub/tags/1.0.0) (2019-11-08)

**Added:**
* `PIVEAU_LOG_LEVEL` for general log level configuration of the `io.piveau` package
* `PIVEAU_HUB_ADDRESS`, `PIVEAU_HUB_APIKEY` in case of missing endpoint in pipe segment config
* `PIVEAU_HUB_ADD_HASH` to add canonical hash if no hash is present
* Ability to export only the Datasets Metrics to the Hub

**Changed:**
* Requires now latest LTS Java 11
* Docker base image to openjdk:11-jre

**Fixed:**
* Update all dependencies

## [0.1.2](https://gitlab.fokus.fraunhofer.de/viaduct/piveau-exporting-hub/tags/0.1.2) (2019-06-21)

**Added:**
* buildInfo.json for build info via `/health` path
* config.schema.json

**Changed:**
* `PIVEAU_` prefix to logstash configuration environment variables

**Fixed:**
* encode identifier when delete

## [0.1.1](https://gitlab.fokus.fraunhofer.de/viaduct/piveau-exporting-hub/tags/0.1.1) (2019-05-17)

**Fixed:**
* Hub address structure reading in delete

## [0.1.0](https://gitlab.fokus.fraunhofer.de/viaduct/piveau-exporting-hub/tags/0.1.0) (2019-05-17)

**Changed:**
* Configuration structure of pipe segment configuration
* Readme

## [0.0.1](https://gitlab.fokus.fraunhofer.de/viaduct/piveau-exporting-hub/tags/0.0.1) (2019-05-03)
Initial release


src/test/java/io/piveau/exporting/ExportingTest.java
package io.piveau.exporting;

import io.vertx.core.Vertx;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.client.WebClient;
import io.vertx.junit5.Checkpoint;
import io.vertx.junit5.Timeout;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("Testing the exporter")
@ExtendWith(VertxExtension.class)
class ExportingTest {

    @BeforeEach
    void startImporter(Vertx vertx, VertxTestContext testContext) {
        vertx.deployVerticle(new MainVerticle(), testContext.completing());
    }

    @Test
    @DisplayName("Receiving dataset and exporting it to hub")
    @Timeout(value = 1, timeUnit = TimeUnit.MINUTES)
    void sendDataset(Vertx vertx, VertxTestContext testContext) {
        Checkpoint checkpoint = testContext.checkpoint(2);

        // Mockup hub
        vertx.createHttpServer().requestHandler(request -> {
            testContext.verify(() -> {
                assertEquals("any-kind", request.getParam("catalogue"));
                assertEquals("text/plain", request.getHeader("Content-Type"));
                assertEquals("not-necessary", request.getHeader("Authorization"));
                assertTrue(request.path().endsWith("/datasets/whatever"));
                request.bodyHandler(buffer -> {
                    assertEquals("Test message", buffer.toString());
                });
            });
            request.response().setStatusCode(200).end(ar -> {
                if(ar.succeeded()) {
                    checkpoint.flag();
                } else {
                    testContext.failNow(ar.cause());
                }
            });
        }).listen(8098);

        // Injecting pipe
        sendPipe("test1-pipe.json", vertx, testContext, checkpoint);
    }

    @Test
    @DisplayName("Receiving identifier list and deleting one dataset")
    @Timeout(value = 1, timeUnit = TimeUnit.MINUTES)
    void sendIdentifierList(Vertx vertx, VertxTestContext testContext) {
        Checkpoint checkpoint = testContext.checkpoint(3);

        // Mockup hub
        vertx.createHttpServer().requestHandler(request -> {
            if (request.method() == HttpMethod.GET && request.path().equals("/hub/datasets")) {
                // assert fetch identifiers
                testContext.verify(() -> {
                    assertEquals("any-kind", request.getParam("catalogue"));
                    assertEquals("not-necessary", request.getHeader("Authorization"));
                });
                request.response().setStatusCode(200).end(new JsonArray().add("first-identifier").add("second-identifier").toBuffer(), ar -> {
                    if(ar.succeeded()) {
                        checkpoint.flag();
                    } else {
                        testContext.failNow(ar.cause());
                    }
                });
            } else if (request.method() == HttpMethod.DELETE && request.path().equals("/hub/datasets/second-identifier")) {
                // assert delete
                testContext.verify(() -> {
                    assertEquals("any-kind", request.getParam("catalogue"));
                    assertEquals("not-necessary", request.getHeader("Authorization"));
                });
                request.response().setStatusCode(200).end(ar -> {
                    if(ar.succeeded()) {
                        checkpoint.flag();
                    } else {
                        testContext.failNow(ar.cause());
                    }
                });
            } else {
                testContext.failNow(new Throwable("Unexpected request to hub."));
            }
        }).listen(8098);

        // Injecting pipe
        sendPipe("test2-pipe.json", vertx, testContext, checkpoint);
    }


    @Test
    @DisplayName("Receiving dataset with metrics and exporting the metrics to hub")
    @Timeout(value = 1, timeUnit = TimeUnit.MINUTES)
    void sendMetrics(Vertx vertx, VertxTestContext testContext) {
        Checkpoint checkpoint = testContext.checkpoint(2);

        // Mockup hub
        vertx.createHttpServer().requestHandler(request -> {
            testContext.verify(() -> {
                assertEquals("any-kind", request.getParam("catalogue"));
                assertEquals("application/trig", request.getHeader("Content-Type"));
                assertEquals("not-necessary", request.getHeader("Authorization"));
                assertTrue(request.path().endsWith("/metrics/sampleDataset"));
            });
            request.response().setStatusCode(200).end(ar -> {
                if (ar.succeeded()) {
                    checkpoint.flag();
                } else {
                    testContext.failNow(ar.cause());
                }
            });
        }).listen(8098);

        // Injecting pipe
        sendPipe("test3-pipe.json", vertx, testContext, checkpoint);
    }

    private void sendPipe(String pipeFile, Vertx vertx, VertxTestContext testContext, Checkpoint checkpoint) {
        vertx.fileSystem().readFile("src/test/resources/" + pipeFile, result -> {
            if (result.succeeded()) {
                JsonObject pipe = new JsonObject(result.result());
                WebClient client = WebClient.create(vertx);
                client.post(8080, "localhost", "/pipe")
                        .putHeader("Content-Type", "application/json")
                        .sendJsonObject(pipe, testContext.succeeding(response -> testContext.verify(() -> {
                            if (response.statusCode() == 202) {
                                checkpoint.flag();
                            } else {
                                testContext.failNow(new Throwable(response.statusMessage()));
                            }
                        })));
            } else {
                testContext.failNow(result.cause());
            }
        });
    }

}


src/main/java/io/piveau/exporting/MainVerticle.java
package io.piveau.exporting;

import io.piveau.exporting.hub.ExportingHubVerticle;
import io.piveau.pipe.connector.PipeConnector;
import io.vertx.core.*;

import java.util.Arrays;

public class MainVerticle extends AbstractVerticle {

    @Override
    public void start(Promise<Void> startPromise) {
        vertx.deployVerticle(ExportingHubVerticle.class, new DeploymentOptions().setWorker(true), ar -> {
            if (ar.succeeded()) {
                PipeConnector.create(vertx, cr -> {
                    if (cr.succeeded()) {
                        cr.result().consumerAddress(ExportingHubVerticle.ADDRESS);
                        startPromise.complete();
                    } else {
                        startPromise.fail(cr.cause());
                    }
                });
            } else {
                startPromise.fail(ar.cause());
            }
        });
    }

    public static void main(String[] args) {
        String[] params = Arrays.copyOf(args, args.length + 1);
        params[params.length - 1] = MainVerticle.class.getName();
        Launcher.executeCommand("run", params);
    }

}


src/main/java/io/piveau/exporting/hub/ExportingHubVerticle.java
package io.piveau.exporting.hub;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.piveau.pipe.Pipe;
import io.piveau.pipe.connector.PipeContext;
import io.piveau.utils.JenaUtils;
import io.vertx.config.ConfigRetriever;
import io.vertx.config.ConfigRetrieverOptions;
import io.vertx.config.ConfigStoreOptions;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.Promise;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.eventbus.Message;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.client.HttpRequest;
import io.vertx.ext.web.client.HttpResponse;
import io.vertx.ext.web.client.WebClient;
import org.apache.jena.query.Dataset;
import org.apache.jena.rdf.model.Model;
import org.apache.jena.riot.Lang;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Set;
import java.util.stream.Collectors;

public class ExportingHubVerticle extends AbstractVerticle {

    public static final String ADDRESS = "io.piveau.pipe.exporting.hub.queue";

    private Logger log = LoggerFactory.getLogger(getClass());

    private WebClient client;

    private String hubAddress;
    private String hubApiKey;
    private boolean hubAddHash;

    @Override
    public void start(Promise<Void> startPromise) {
        vertx.eventBus().consumer(ADDRESS, this::handlePipe);
        client = WebClient.create(vertx);

        ConfigStoreOptions envStoreOptions = new ConfigStoreOptions()
                .setType("env")
                .setConfig(new JsonObject().put("keys", new JsonArray()
                        .add("PIVEAU_HUB_ADDRESS")
                        .add("PIVEAU_HUB_APIKEY")
                        .add("PIVEAU_HUB_ADD_HASH")));
        ConfigRetriever retriever = ConfigRetriever.create(vertx, new ConfigRetrieverOptions().addStore(envStoreOptions));
        retriever.getConfig(ar -> {
            if (ar.succeeded()) {
                hubAddress = ar.result().getString("PIVEAU_HUB_ADDRESS", "http://piveau-hub:8080");
                hubApiKey = ar.result().getString("PIVEAU_HUB_APIKEY", "");
                hubAddHash = ar.result().getBoolean("PIVEAU_HUB_ADD_HASH", false);
                startPromise.complete();
            } else {
                startPromise.fail(ar.cause());
            }
        });
    }

    private boolean fieldEquals(ObjectNode dataInfo, String field, String value) {
        return dataInfo.hasNonNull(field) && dataInfo.get(field).asText().equals(value);
    }

    private void handlePipe(Message<PipeContext> message) {
        PipeContext pipeContext = message.body();
        Pipe pipe = pipeContext.getPipe();
        pipeContext.log().debug(Json.encodePrettily(pipe));

        ObjectNode dataInfo = pipeContext.getDataInfo();

        JsonNode config = pipeContext.getConfig();

        ObjectNode hub = ((ObjectNode)config).with("hub");
        String address = hub.path("endpoint").path("address").asText(hubAddress);
        String apiKey = hub.path("endpoint").path("apiKey").asText(hubApiKey);

        if(fieldEquals(dataInfo, "content", "identifierList")) {
            deleteIdentifiers(address, apiKey, pipeContext);
        } else if(fieldEquals(dataInfo, "content", "metrics")) {
            exportMetrics(address, apiKey, pipeContext);
        } else {
            exportMetadata(address, apiKey, pipeContext);
        }
    }

    private void exportMetadata(String address, String apiKey, PipeContext pipeContext) {

        JsonNode dataInfo = pipeContext.getDataInfo();

        String identifier = URLEncoder.encode(dataInfo.path("identifier").asText(), StandardCharsets.UTF_8);

        log.debug("Content type: {}", pipeContext.getMimeType());

        HttpRequest<Buffer> request = client.putAbs(address + "/datasets/" + identifier)
                .putHeader("Authorization", apiKey)
                .putHeader("Content-Type", pipeContext.getMimeType())
                .addQueryParam("catalogue", dataInfo.path("catalogue").textValue());

        if (dataInfo.hasNonNull("hash")) {
            request.addQueryParam("hash", dataInfo.path("hash").textValue());
        } else if (hubAddHash) {
            Model model = JenaUtils.read(pipeContext.getStringData().getBytes(), pipeContext.getMimeType());
            request.addQueryParam("hash", JenaUtils.canonicalHash(model));
        }

        request.sendBuffer(Buffer.buffer(pipeContext.getStringData()), ar -> {
            if (ar.succeeded()) {
                if (ar.result().statusCode() == 201) {
                    pipeContext.log().info("Dataset created: {}", dataInfo);
                } else if (ar.result().statusCode() == 200) {
                    pipeContext.log().info("Dataset updated: {}", dataInfo);
                } else if (ar.result().statusCode() == 304) {
                    pipeContext.log().info("Dataset skipped: {}", dataInfo);
                } else {
                    pipeContext.setFailure(new Throwable(identifier, new Throwable("" + ar.result().statusCode() + " - " + ar.result().statusMessage() + " - " + ar.result().bodyAsString())));
                }
            } else {
                pipeContext.setFailure(ar.cause());
            }
        });
    }

    private void exportMetrics(String address, String apiKey, PipeContext pipeContext) {

        JsonNode dataInfo = pipeContext.getDataInfo();

        String identifier = URLEncoder.encode(dataInfo.path("identifier").asText(), StandardCharsets.UTF_8);

        Dataset ds = JenaUtils.readDataset(pipeContext.getStringData().getBytes(),null);

        HttpRequest<Buffer> request = client.putAbs(address + "/metrics/" + identifier)
                .putHeader("Authorization", apiKey)
                .putHeader("Content-Type", pipeContext.getMimeType())
                .addQueryParam("catalogue", dataInfo.path("catalogue").textValue());

        request.sendBuffer(Buffer.buffer(pipeContext.getStringData()), ar -> {
            if(ar.succeeded()) {
                if(ar.result().statusCode() == 201) {
                    pipeContext.log().info("Metrics graph created: {}", dataInfo);
                } else if(ar.result().statusCode() == 200) {
                    pipeContext.log().info("Metrics graph updated: {}", dataInfo);
                }  else {
                    pipeContext.setFailure(new Throwable(identifier, new Throwable("" + ar.result().statusCode() + " - " + ar.result().statusMessage() + " - " + ar.result().bodyAsString())));
                }
            } else {
                pipeContext.setFailure(ar.cause());
            }
        });
    }


    private void deleteIdentifiers(String address, String apiKey, PipeContext pipeContext) {

        String catalogueId = pipeContext.getDataInfo().path("catalogue").asText();
        HttpRequest<Buffer> request = client.getAbs(address + "/datasets")
                .putHeader("Authorization", apiKey)
                .addQueryParam("catalogue", catalogueId)
                .addQueryParam("sourceIds", "true");

        request.send(ar -> {
            if (ar.succeeded()) {
                HttpResponse<Buffer> response = ar.result();
                if (response.statusCode() == 200) {
                    Set<String> sourceIds = new JsonArray(pipeContext.getStringData()).stream().map(Object::toString).collect(Collectors.toSet());
                    Set<String> targetIds = ar.result().bodyAsJsonArray().stream().map(Object::toString).collect(Collectors.toSet());
                    int targetSize = targetIds.size();
                    targetIds.removeAll(sourceIds);
                    pipeContext.log().info("Source {}, target {}, deleting {} datasets", sourceIds.size(), targetSize, targetIds.size());
                    targetIds.forEach(datasetId -> deleteDataset(address, apiKey, pipeContext, datasetId, catalogueId));
                } else {
                    pipeContext.log().error(response.statusMessage());
                }
            } else {
                pipeContext.setFailure(ar.cause());
            }
        });
    }

    private void deleteDataset(String address, String apiKey, PipeContext pipeContext, String datasetId, String catalogueId) {
        String encoded = URLEncoder.encode(datasetId, StandardCharsets.UTF_8);

        HttpRequest<Buffer> request = client.deleteAbs(address + "/datasets/" + encoded)
                .putHeader("Authorization", apiKey)
                .addQueryParam("catalogue", catalogueId);
        request.send(ar -> {
            if (ar.succeeded()) {
                HttpResponse<Buffer> response = ar.result();
                switch (response.statusCode()) {
                    case 200:
                        pipeContext.log().info("Dataset '{}' deleted", datasetId);
                        break;
                    case 404:
                        pipeContext.log().warn("Dataset '{}' not found", datasetId);
                        break;
                    default:
                        pipeContext.log().error("{} - {} ({})", response.statusCode(), response.statusMessage(), datasetId);
                }
            } else {
                pipeContext.log().error("Delete dataset", ar.cause());
            }
        });
    }

}
